guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+Even better, however, is to make use of [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) to eliminate the need for a global. This allows you to create a fake version of the value in question within your test:
+
+2️⃣
+
+``` typescript
+describe("interpretMethodData", () => {
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    const provider = createFakeProvider({ network: "sepolia" });
+    expect(interpretMethodData(provider, '0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    const provider = createFakeProvider({ network: "mainnet" });
+    expect(interpretMethodData(provider, '0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+### Shared variables
+
+If a variable used inside of a test is declared outside of the test, the value that the variable holds — and thus changes to that variable made inside the test — will remain even after the test ends.
+
+🚫
+
+``` typescript
+const NETWORK = {
+  provider: new HttpProvider('https://mainnet.infura.io/v3/abc123');
+};
+
+describe("interpretMethodData", () => {
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    // This change will apply to every other test that's run after this one
+    NETWORK.provider = new HttpProvider(
+      'https://sepolia.infura.io/v3/abc123',
+    );
+    expect(interpretMethodData('0x3fbac0ab', NETWORK)).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    // Oops! This will still hit Sepolia
+    expect(interpretMethodData('0x3fbac0ab', NETWORK)).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+It is tempting to reach for `beforeEach` to correct this:
+
+🚫
+
+``` typescript
+describe("interpretMethodData", () => {
+  const network;
+
+  beforeEach(() => {
+    // Now this variable is reset before each test, creating a clean slate
+    network = {
+      provider: new HttpProvider('https://mainnet.infura.io/v3/abc123');
+    };
+  });
+
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    // This change will no longer apply to every other test
+    network.provider = new HttpProvider(
+      'https://sepolia.infura.io/v3/abc123',
+    );
+    expect(interpretMethodData('0x3fbac0ab', network)).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    // Okay, this will still use Mainnet
+    expect(interpretMethodData('0x3fbac0ab', network)).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+However, it is best to [avoid hooks altogether](#avoid-before-each-and-after-each). Instead, use a factory function which defines defaults and allows them to be overridden where they are needed:
```

Comment:
Nit: Slightly softer wording here might be better
```suggestion
However, it is recommended to [avoid hooks altogether](#avoid-before-each-and-after-each). Instead, use a factory function which defines defaults and allows them to be overridden where they are needed:
```

Reply:
In another subsection I used "Instead of using hooks, ..." so I copied that wording to here to be consistent. Does that phrase still need revision?

Reply:
That sounds good too! Just wanted to avoid "best"